### DISCUSSION TITLE
Enhanced API for @astrojs/vue's `appEntryPoint`

### DISCUSSION THREAD

**Discussion by rbsummers** [withastro/roadmap#810] [Proposal]
# Summary

I'd like to request expanding the information that the [`appEntryPoint`](https://docs.astro.build/en/guides/integrations-guide/vue/#appentrypoint) feature of the Vue integration receives when dealing with server side rendered requests. Having access to the `Astro` global would be ideal, but just the request or the `Astro.locals` context would be a good start already.

# Background & Motivation

I'm a developer of a new Astro app that will have some `*.astro` pages where both Astro and Vue components will be used together. Of those on the second group, some are already in place and used elsewhere, also we want to selectively use client or server rendering for them. Existing components use `vue-i18n` and we want to keep using that for the new ones.

The language selection is implemented using a middleware that will pick the language from the search params, and we store this information in the `Astro.locals` context in order to be used further down the line. This works fine and I can access the data in any `*.astro` page.

But the issue is that I haven't found a way to make this specific request-bound language value (or any other kind of request information) available when instancing the Vue app, and given that each Vue component gets [it's own application instance](https://github.com/withastro/astro/issues/6312), any attempts to make a wrapper / high order component that changes this globally fails, as the global `i18n` instance it's not the same across components. I could keep a single instance of the `i18n` plugin and reuse that for every application, but this will create a problem for server rendered components as handling requests from different users with different selected languages concurrently will cause incorrect output.

My only option left is to specifically add a `language` attribute to every Vue component and explicitly pass it down when instantiating them on any `*.astro` page, but I dislike this as this will pollute every component with something they shouldn't really have to worry about as the global language should be an app concern.

For client side components I've achieved some solutions easily by either using an `atom` from the `nanostores` lib and using `import.meta.env.SSR` to identify when I can access `window.location.href`. Although this works, an unified client and server side render API would be better if possible.

# Goals

* Make more information available to the underlying Vue apps on the `appEntryPoint` "hook". The guide mentions that this is intended for advanced use cases but the data available here falls short for some scenarios

# Example

```
export default (app: App, astroGlobal: Astro) => {
  app.use(
    createI18n({
      // Where previously we didn't have any way of choosing this using request data
      // locale: 'en',

      // Now we can, for example, do
      locale: astroGlobal.locals.language
    })
  );
}
```

# Alternatives

On [#6312](https://github.com/withastro/astro/issues/6312) it was proposed that if each component was going to have it's own app, the component should also be able to specify how it will be created, and I think that is worth considering for even more complex use cases. Maybe introducing an specific tag that will mark the scope of the created app & it's setup method in order to achieve more fine grained control? Something along these lines:

```
const setupFn: (app: App) => {
  // ...
}
---
<AstroVueApp setup={setupFn}>
  <VueComponent1>
    <AstroComponent />
  </VueComponent1>

  <VueComponent2 />
</AstroVueApp>
```

**Comment by leomp12** [comment]
From what I tested, current `@astrojs/vue/server.js` implementation full `Astro` global is not available, but [`Astro.cookies`](https://docs.astro.build/en/reference/api-reference/#astrocookies) is available with `this.result?.cookies`, an example I'm using:

```js
const sid = this.result?.cookies?.get('sid')?.value;
// ...
app.provide('sid', sid);
```

```js
// server.js
import { h, createSSRApp } from 'vue';
import { renderToString } from 'vue/server-renderer';
import { setup } from 'virtual:@astrojs/vue/app';
import StaticHtml from './static-html.js';

function check(Component) {
	return !!Component['ssrRender'] || !!Component['__ssrInlineRender'];
}

async function renderToStaticMarkup(Component, inputProps, slotted, metadata) {
	const slots = {};
	const props = { ...inputProps };
	delete props.slot;
	const sid = this.result?.cookies?.get('sid')?.value;
	for (const [key, value] of Object.entries(slotted)) {
		slots[key] = () =>
			h(StaticHtml, {
				value,
				name: key === 'default' ? undefined : key,
				// Adjust how this is hydrated only when the version of Astro supports `astroStaticSlot`
				hydrate: metadata.astroStaticSlot ? !!metadata.hydrate : true,
			});
	}
	const app = createSSRApp({ render: () => h(Component, props, slots) });
	app.provide('sid', sid);
	await setup(app, { sid });
	const html = await renderToString(app);
	return { html };
}

export default {
	check,
	renderToStaticMarkup,
	supportsAstroStaticSlot: true,
};
```

**Reply by leomp12** [reply]
I mean, the best scenario would be to be able to have `Astro.locals` in the Vue instance IMO, but if this is unfeasible, having server-side cookies from the respective execution is enough for my case and I believe for your example with `lang` as well, and it would be a simpler change apparently

**Comment by kindlyfire** [comment]
I'd like to request something similar. My use-case is that I need the current page URL to render the right page during SSR. My problem was solved through using [astro-als](https://github.com/florian-lefebvre/astro-als). I'd love something similar landing in Astro itself.

Ideally we'd have access to `Astro` during SSR. I don't need it client-side per se
